client/src/client.rs

```diff
@@ -432,30 +440,42 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		storage_key: Option<&StorageKey>,
 		key: &StorageKey
 	) -> sp_blockchain::Result<Vec<(NumberFor<Block>, u32)>> {
-		let (config, storage) = self.require_changes_trie()?;
 		let last_number = self.backend.blockchain().expect_block_number_from_id(&last)?;
 		let last_hash = self.backend.blockchain().expect_block_hash_from_id(&last)?;
+		let (storage, configs) = self.require_changes_trie(first, last_hash, false)?;
+
+		let mut result = Vec::new();
+		let best_number = self.backend.blockchain().info().best_number;
+		for (config_zero, config_end, config) in configs {
+			let range_first = ::std::cmp::max(first, config_zero + One::one());
+			let range_anchor = match config_end {
+				Some((config_end_number, config_end_hash)) => if last_number > config_end_number {
+					ChangesTrieAnchorBlockId { hash: config_end_hash, number: config_end_number }
+				} else {
+					ChangesTrieAnchorBlockId { hash: convert_hash(&last_hash), number: last_number }
+				},
+				None => ChangesTrieAnchorBlockId { hash: convert_hash(&last_hash), number: last_number },
+			};
 
-		// FIXME: remove this in https://github.com/paritytech/substrate/pull/3201
-		let config_range = ChangesTrieConfigurationRange {
-			config: &config,
-			zero: Zero::zero(),
-			end: None,
-		};
+			let config_range = ChangesTrieConfigurationRange {
+				config: &config,
+				zero: config_zero.clone(),
+				end: config_end.map(|(config_end_number, _)| config_end_number),
+			};
+			let result_range: Vec<(NumberFor<Block>, u32)> = key_changes::<Blake2Hasher, _>(
+				config_range,
+				storage.storage(),
+				range_first,
+				&range_anchor,
+				best_number,
+				storage_key.as_ref().map(|x| &x.0[..]),
+				&key.0)
+			.and_then(|r| r.map(|r| r.map(|(block, tx)| (block, tx))).collect::<Result<_, _>>())
+			.map_err(|err| sp_blockchain::Error::ChangesTrieAccessFailed(err))?;
+			result.extend(result_range);
```

Comment:
result_range is descending order, trie configs are descending order, so I would rather do
`result = result_range.extend(result)`, which is inefficient but return globally descending order.
(edit: I may be mistaken on order, conclusion is just from reading doc of key_changes and require_changes_trie).

Reply:
Not sure - first we call `require_changes_trie`, which e.g. returns `vec![(50..100), (30..50)]`. So we start with the range 50..100, for which we call `key_changes`, which e.g. first returns changes for block 100, then for 50. Then we'll continue with range 30..50 => result is extended with `vec![changes_for_block_50, changes_for_block_30]`. Everything looks ok to me. Am I wrong?

Reply:
You are totally right, for whatever reason I expected this method to return result in ascending order when it is written in the doc that it will return descending (and it got the same name as the other documented descending method).